clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -209,20 +210,27 @@ public CompletableFuture start() throws Exception {
         }
 
         transport.setOnReceive(this.callback);
-        return transport.start().thenCompose((future) -> {
+        return CompletableFuture.allOf(handshakeResposeFuture, transport.start().thenCompose((future) -> {
             String handshake = HandshakeProtocol.createHandshakeRequestMessage(new HandshakeRequestMessage(protocol.getName(), protocol.getVersion()));
-            return transport.send(handshake).thenRun(() -> {
-                hubConnectionStateLock.lock();
-                try {
-                    hubConnectionState = HubConnectionState.CONNECTED;
-                    connectionState = new ConnectionState(this);
-                    logger.log(LogLevel.Information, "HubConnected started.");
-                } finally {
-                    hubConnectionStateLock.unlock();
-                }
-            });
-        });
+            return transport.send(handshake);
+        })).thenRun(() -> {
+            hubConnectionStateLock.lock();
+            try {
+                hubConnectionState = HubConnectionState.CONNECTED;
+                connectionState = new ConnectionState(this);
+                logger.log(LogLevel.Information, "HubConnected started.");
+            } finally {
+                hubConnectionStateLock.unlock();
+            }
+        }).acceptEither(timeoutAfter(5, TimeUnit.SECONDS), startFuture -> {});
```

Comment:
I don't think we want a timeout as the server is the one who sends a handshake response with an error if it times out.

And if the connection to the server is dropped then we should `completeExceptionally` the `handshakeResponseFuture` in the `stop` function. Which it looks like we don't wire up to the transport yet.

Reply:
So looks like we actually will need the timeout, and we'll have to set it to longer than what the server timeout is. If we sent the handshake and the server went offline we could potentially wait for the response forever